Error messages could be improved

This demonstrates the issue: https://play.golang.org/p/SwaihNVSqAl

I personally would be fine with the field being set to `a` or `A`, but `a.` is pretty clearly wrong.

So I made a little test to compare the values of various mistyped parses to compare `encoding/json` and this lib.  Here are the results:

```
--- FAIL: TestDecodeErrors (0.00s)
    --- FAIL: TestDecodeErrors/0 (0.00s)
        errors_test.go:70: value doesn't match encoding/json, got "\"1}\"", expected "number"
        errors_test.go:73: offset doesn't match encoding/json, got 0, expected 9
        errors_test.go:76: struct doesn't match encoding/json, got "json.example", expected "example"
        errors_test.go:79: field incorrect, got "name.", expected "Name"
        errors_test.go:82: representation doesn't match encoding/json, got "json: cannot unmarshal \"1}\" into Go struct field json.example.name. of type string", expected "json: cannot unmarshal number into Go struct field example.Name of type string"
    --- FAIL: TestDecodeErrors/1 (0.00s)
        errors_test.go:70: value doesn't match encoding/json, got "\"1}}\"", expected "number"
        errors_test.go:73: offset doesn't match encoding/json, got 0, expected 24
        errors_test.go:76: struct doesn't match encoding/json, got "json.examplejson.Inner", expected "Inner"
        errors_test.go:79: field incorrect, got "inner.innername.", expected "Inner.InnerName"
        errors_test.go:82: representation doesn't match encoding/json, got "json: cannot unmarshal \"1}}\" into Go struct field json.examplejson.Inner.inner.innername. of type string", expected "json: cannot unmarshal number into Go struct field Inner.Inner.InnerName of type string"
    --- FAIL: TestDecodeErrors/2 (0.00s)
        errors_test.go:70: value doesn't match encoding/json, got "\"1}\"", expected "number"
        errors_test.go:73: offset doesn't match encoding/json, got 0, expected 8
        errors_test.go:76: struct doesn't match encoding/json, got "json.example", expected "example"
        errors_test.go:79: field incorrect, got "map.", expected "Map"
        errors_test.go:82: representation doesn't match encoding/json, got "json: cannot unmarshal \"1}\" into Go struct field json.example.map. of type map[string]json.MapStruct", expected "json: cannot unmarshal number into Go struct field example.Map of type map[string]json.MapStruct"
    --- FAIL: TestDecodeErrors/3 (0.00s)
        errors_test.go:70: value doesn't match encoding/json, got "\"1}}\"", expected "number"
        errors_test.go:73: offset doesn't match encoding/json, got 0, expected 19
        errors_test.go:76: struct doesn't match encoding/json, got "json.examplemap[string]json.MapStruct{}", expected "example"
        errors_test.go:79: field incorrect, got "map.mapname.", expected "Map"
        errors_test.go:82: representation doesn't match encoding/json, got "json: cannot unmarshal \"1}}\" into Go struct field json.examplemap[string]json.MapStruct{}.map.mapname. of type json.MapStruct", expected "json: cannot unmarshal number into Go struct field example.Map of type json.MapStruct"
    --- FAIL: TestDecodeErrors/4 (0.00s)
        errors_test.go:70: value doesn't match encoding/json, got "\"1}\"", expected "number"
        errors_test.go:73: offset doesn't match encoding/json, got 0, expected 10
        errors_test.go:76: struct doesn't match encoding/json, got "json.example", expected "example"
        errors_test.go:79: field incorrect, got "slice.", expected "Slice"
        errors_test.go:82: representation doesn't match encoding/json, got "json: cannot unmarshal \"1}\" into Go struct field json.example.slice. of type []json.SliceStruct", expected "json: cannot unmarshal number into Go struct field example.Slice of type []json.SliceStruct"
    --- FAIL: TestDecodeErrors/5 (0.00s)
        errors_test.go:70: value doesn't match encoding/json, got "\"1}]}\"", expected "number"
        errors_test.go:73: offset doesn't match encoding/json, got 0, expected 24
        errors_test.go:76: struct doesn't match encoding/json, got "json.example[]json.SliceStructjson.SliceStruct", expected "SliceStruct"
        errors_test.go:79: field incorrect, got "slice.0.slicename.", expected "Slice.SliceName"
        errors_test.go:82: representation doesn't match encoding/json, got "json: cannot unmarshal \"1}]}\" into Go struct field json.example[]json.SliceStructjson.SliceStruct.slice.0.slicename. of type string", expected "json: cannot unmarshal number into Go struct field SliceStruct.Slice.SliceName of type string"
```

The only thing I intended to fix was the Field value, though I think Field and Struct are clearly related so I would be willing to call them both worth fixing.  Here are the issues with `encoding/json`:

 * error context is not as good, just says the type that was wrong, not a fragment like this lib
 * no extra context on nested types; the inclusion of the package and enclosing types by this library is clearly useful

Here are the issues with `github.com/segmentio/encoding`:

 * All fields get a silly trailing dot.  Likely easy to fix.
 * fields are in terms of json, not go fieldnames.  I suspect this is easy to fix.
 * complex types end up messy or inscrutable (eg `json.example[]json.SliceStructjson.SliceStruct`)

---

My plan is to remove the trailing dot, correct fields to match Go fields, not json fields, and *maybe* clean up the complex types.  The only thing that makes me nervous about this is that if anyone is depending on these strings (I am sure some people do) their error checking code will silently stop working.  I would be willing to add a new field or fields to the error struct but that would imply no longer making a type alias, which I'm sure would break yet more code.

Is it worth fixing this and causing a break in compatibility?

---

(and for posterity, the actual test:

```golang
package json

import (
	"encoding/json"
	"errors"
	"strconv"
	"strings"
	"testing"
)

func TestDecodeErrors(t *testing.T) {
	type Test struct{ in, err string }
	type Inner struct {
		InnerName string
	}
	type MapStruct struct {
		MapName string
	}
	type SliceStruct struct {
		SliceName string
	}
	type example struct {
		Name string

		Inner Inner

		Map map[string]MapStruct

		Slice []SliceStruct
	}
	tests := []Test{
		0: {`{"name":1}`, ""},
		1: {`{"inner":{"innername": 1}}`, ""},

		2: {`{"map":1}`, ""},
		3: {`{"map":{"mapname":1}}`, ""},

		4: {`{"slice":1}`, ""},
		5: {`{"slice":[{"slicename":1}]}`, ""},
	}

	for i, tt := range tests {
		t.Run(strconv.Itoa(i), func(t *testing.T) {
			var dest example
			d := NewDecoder(strings.NewReader(tt.in))
			err := d.Decode(&dest)
			if err == nil {
				t.Error("expected an error but got none")
				return
			}
			typedErr := &UnmarshalTypeError{}
			if !errors.As(err, &typedErr) {
				t.Errorf("expected UnmarshalTypeError, got %T", err)
				return
			}
			d2 := json.NewDecoder(strings.NewReader(tt.in))
			expectedErr := d2.Decode(&dest)
			if expectedErr == nil {
				t.Error("expected an error but got none")
				return
			}

			typedExpectedErr := &json.UnmarshalTypeError{}
			if !errors.As(expectedErr, &typedExpectedErr) {
				t.Errorf("expected UnmarshalTypeError, got %T", expectedErr)
				return
			}

			if typedErr.Value != typedExpectedErr.Value {
				t.Logf("value doesn't match encoding/json, got %q, expected %q", typedErr.Value, typedExpectedErr.Value)
			}
			if typedErr.Offset != typedExpectedErr.Offset {
				t.Logf("offset doesn't match encoding/json, got %d, expected %d", typedErr.Offset, typedExpectedErr.Offset)
			}
			if typedErr.Struct != typedExpectedErr.Struct {
				t.Logf("struct doesn't match encoding/json, got %q, expected %q", typedErr.Struct, typedExpectedErr.Struct)
			}
			if typedErr.Field != typedExpectedErr.Field {
				t.Errorf("field incorrect, got %q, expected %q", typedErr.Field, typedExpectedErr.Field)
			}
			if typedErr.Error() != typedExpectedErr.Error() {
				t.Logf("representation doesn't match encoding/json, got %q, expected %q", typedErr, typedExpectedErr)
			}
		})
	}
}
```

--- a/json/decode.go
+++ b/json/decode.go
@@ -535,7 +535,7 @@ func (d decoder) decodeArray(b []byte, p unsafe.Pointer, n int, size uintptr, t 
 		if err != nil {
 			if e, ok := err.(*UnmarshalTypeError); ok {
 				e.Struct = t.String() + e.Struct
-				e.Field = strconv.Itoa(i) + "." + e.Field
+				e.Field = d.prependField(strconv.Itoa(i), e.Field)
 			}
 			return b, err
 		}
@@ -637,7 +637,7 @@ func (d decoder) decodeSlice(b []byte, p unsafe.Pointer, size uintptr, t reflect
 			}
 			if e, ok := err.(*UnmarshalTypeError); ok {
 				e.Struct = t.String() + e.Struct
-				e.Field = strconv.Itoa(s.len) + "." + e.Field
+				e.Field = d.prependField(strconv.Itoa(s.len), e.Field)
 			}
 			return b, err
 		}
@@ -716,7 +716,7 @@ func (d decoder) decodeMap(b []byte, p unsafe.Pointer, t, kt, vt reflect.Type, k
 			}
 			if e, ok := err.(*UnmarshalTypeError); ok {
 				e.Struct = "map[" + kt.String() + "]" + vt.String() + "{" + e.Struct + "}"
-				e.Field = fmt.Sprint(k.Interface()) + "." + e.Field
+				e.Field = d.prependField(fmt.Sprint(k.Interface()), e.Field)
 			}
 			return b, err
 		}
@@ -797,7 +797,7 @@ func (d decoder) decodeMapStringInterface(b []byte, p unsafe.Pointer) ([]byte, e
 			}
 			if e, ok := err.(*UnmarshalTypeError); ok {
 				e.Struct = mapStringInterfaceType.String() + e.Struct
-				e.Field = key + "." + e.Field
+				e.Field = d.prependField(key, e.Field)
 			}
 			return b, err
 		}
@@ -878,7 +878,7 @@ func (d decoder) decodeMapStringRawMessage(b []byte, p unsafe.Pointer) ([]byte, 
 			}
 			if e, ok := err.(*UnmarshalTypeError); ok {
 				e.Struct = mapStringRawMessageType.String() + e.Struct
-				e.Field = key + "." + e.Field
+				e.Field = d.prependField(key, e.Field)
 			}
 			return b, err
 		}
@@ -959,7 +959,7 @@ func (d decoder) decodeMapStringString(b []byte, p unsafe.Pointer) ([]byte, erro
 			}
 			if e, ok := err.(*UnmarshalTypeError); ok {
 				e.Struct = mapStringStringType.String() + e.Struct
-				e.Field = key + "." + e.Field
+				e.Field = d.prependField(key, e.Field)
 			}
 			return b, err
 		}
@@ -1041,7 +1041,7 @@ func (d decoder) decodeMapStringStringSlice(b []byte, p unsafe.Pointer) ([]byte,
 			}
 			if e, ok := err.(*UnmarshalTypeError); ok {
 				e.Struct = mapStringStringType.String() + e.Struct
-				e.Field = key + "." + e.Field
+				e.Field = d.prependField(key, e.Field)
 			}
 			return b, err
 		}
@@ -1125,7 +1125,7 @@ func (d decoder) decodeMapStringBool(b []byte, p unsafe.Pointer) ([]byte, error)
 			}
 			if e, ok := err.(*UnmarshalTypeError); ok {
 				e.Struct = mapStringStringType.String() + e.Struct
-				e.Field = key + "." + e.Field
+				e.Field = d.prependField(key, e.Field)
 			}
 			return b, err
 		}
@@ -1215,7 +1215,7 @@ func (d decoder) decodeStruct(b []byte, p unsafe.Pointer, st *structType) ([]byt
 			}
 			if e, ok := err.(*UnmarshalTypeError); ok {
 				e.Struct = st.typ.String() + e.Struct
-				e.Field = string(k) + "." + e.Field
+				e.Field = d.prependField(string(k), e.Field)
 			}
 			return b, err
 		}
@@ -1436,4 +1436,12 @@ func (d decoder) decodeTextUnmarshaler(b []byte, p unsafe.Pointer, t reflect.Typ
 	}
 
 	return b, &UnmarshalTypeError{Value: value, Type: reflect.PtrTo(t)}
+}
+
+func (d decoder) prependField(key, field string) string {
+	if field != "" {
+		return key + "." + field
+	} else {
+		return key
+	}
 }

--- a/json/decode.go
+++ b/json/decode.go
@@ -1441,7 +1441,6 @@ func (d decoder) decodeTextUnmarshaler(b []byte, p unsafe.Pointer, t reflect.Typ
 func (d decoder) prependField(key, field string) string {
 	if field != "" {
 		return key + "." + field
-	} else {
-		return key
 	}
+	return key
 }

